sandbox/mooncake.py

```diff
@@ -0,0 +1,252 @@
+#!/bin/python
+
+import numpy as np
+import torch
+from torch import nn, optim
+import torch.nn.functional as F
+import matplotlib.pyplot as plt
+
+DEVICE = "cuda"
+NUM_TRAIN_ITERATIONS = 2000
+SEED_LENGTH = 20
+
+
+class CausalConv1d(nn.Conv1d):
+    # From Alex Rogozhnikov:
+    # https://github.com/pytorch/pytorch/issues/1333#issuecomment-400338207
+    def __init__(
+            self,
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride=1,
+            dilation=1,
+            groups=1,
+            bias=True,
+    ):
+        self.__padding = (kernel_size - 1) * dilation
+        super(CausalConv1d, self).__init__(
+            in_channels,
+            out_channels,
+            kernel_size=kernel_size,
+            stride=stride,
+            padding=self.__padding,
+            dilation=dilation,
+            groups=groups,
+            bias=bias,
+        )
+
+    def forward(self, input):
+        result = super(CausalConv1d, self).forward(input)
+        if self.__padding != 0:
+            return result[:, :, :-self.__padding]
+        return result
+
+
+def generate_sinewaves(batch_size, max_num_samples=200):
+    """
+    Returns samples of one period of a sinewave with random frequency, phase
+    and number of samples.
+
+    Parameters
+    ----------
+    batch_size : int
+        Batch size.
+
+    Returns
+    -------
+    sinewave : [batch_size, 100]
+    """
+    sinewaves = []
+    for i in range(batch_size):
+        n = 100  # FIXME np.random.randint(50, 150)
+        f = 1
+        phi = 0
+        x = np.linspace(0, 2 / f, n, dtype=np.float32)
+        sinewave = np.sin(2 * np.pi * f * (x + phi))
+        pad_length = max_num_samples - len(sinewave)
+        sinewave = np.pad(
+            sinewave, (0, pad_length), "constant", constant_values=0)
+        sinewave += 0.1 * np.random.randn(max_num_samples)
+        sinewaves.append(sinewave)
+    return torch.tensor(np.stack(sinewaves, axis=0)).to(DEVICE)
+
+
+def negative_log_prob(x, pi, mu, sigma):
+    """
+    Negative log probability of predictive distribution of x_{k+1} from x_1,
+    ..., x_k.
+
+    Parameters
+    ----------
+    x : [batch_size, max_num_samples]
+        Sinewave.
+    pi, mu, sigma : [batch_size, num_components (+1 for pi), max_num_samples]
+        Mixture weights, means and standard deviations of each Gaussian
+        component, respectively.
+
+    Returns
+    -------
+    [batch_size, max_num_samples - 1]
+    """
+    # Index appropriately to compute nlogp of predicted next value.
+    vals = x.unsqueeze(1)[..., 1:]
```

Comment:
personally i'd unsqueeze in the data gen func rather than in multiple places in the forward pass/loss calc

Reply:
you can see how i did it in ar-exp